rpc/gnmi/gnmi-history.md

```diff
@@ -0,0 +1,47 @@
+# gNMI History Extension
+
+**Contributors:** Justin Costa-Roberts, Aaron Beitch, Mike Fink
+
+**Date:** August 4, 2020
+
+**Version:** 0.1.0
+
+# 1. Purpose
+
+In certain applications it is necessary not only to retrieve current data from a data tree, but also to allow querying historical data. The gNMI History extension enables clients both to retrieve data at a specific time in the past, and to request all updates applied to a data tree between two specified times.
+
+# 2. Definition
+
+A `History` message is embedded in an Extension message in the `SubscribeRequest` proto.
+
+## 2.1 Proto
+
+```
+message TimeRange {
+  int64 start = 1;
+  int64 end = 2;
+}
+
+message History {
+  oneof request {
+    int64 time = 1;
+    TimeRange range = 2;
+  }
+}
+```
+
+## 2.2 SubscribeRequest handling
+
+The History extension may be applied only to `ONCE` and `STREAM` subscriptions. Requests for historical data can be either of two types: a snapshot at a specified time, or a set of updates between two times. The `History` message's request field is a `oneof` field whose type indicates which of these request types applies.
+
+In all cases, times are represented as int64s interpreted as nanoseconds since the Unix Epoch.
+
+### 2.2.1 Snapshot requests
+
+If the `time` field is set, the request is for a snapshot. The subscription mode for a time request must be `ONCE`; otherwise the server must return an `INVALID_ARGUMENT` error. The server must transmit the data indicated in the `SubscribeRequest` as it existed on the server at the specified time. If the specified time indicates a future time at the point the server receives the request, the server may either return an `UNIMPLEMENTED` error or choose to wait until the specified time has been reached.
```

Comment:
I wonder whether it's a good idea to wait until the specified time has been reached -- it seems like a potential DoS vector if a client can send a lot of requests that potentially are held open for a long time. Is there a specific use case that motivates this?

Reply:
We went back and forth on this. The reason to wait is that it seems undesirable for a client request ending at t=time.Now() to result in an error just because the server's time is slightly behind the client's. Another approach we considered is to have the server immediately return all updates between the specified start time and its own current time u (u < t). This also seems undesirable because the client might well use time t as the start time for a later request (if, say, it does a get every hour), which would create a window between u and t for which it would not receive updates, and wouldn't know it hadn't.

The approach we landed on was either to give the client exactly what it asked for or to return an error, and to leave it up to the server whether it wants to wait. In practice I imagine a server would have some short period it would be willing to wait, and otherwise would return an error.

What do you think?

Reply:
That sounds reasonable to me -- why don't we add a small note here that just explains that a client should expect that there are conditions in which:

* Connections with start times that that are far into the future may be rejected.
* Servers may limit the number of connections of this nature.

And notes that server implementors should be careful not to cause themselves a resource exhaustion problem here. It's stepping beyond a spec into implementation advice, but I think we did something similar in the 'Considerations for using Get' section of the main spec.

Reply:
Yeah that seems good. I've added a section at the end of the document including these points.